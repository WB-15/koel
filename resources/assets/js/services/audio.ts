export const audio = {
  context: null as AudioContext | null,
  source: null as MediaElementAudioSourceNode | null,
  element: null as HTMLMediaElement | null,

  init (element: HTMLMediaElement): void {
    const AudioContext = window.AudioContext ||
      window.webkitAudioContext ||
      window.mozAudioContext ||
      window.oAudioContext ||
      window.msAudioContext

    this.context = new AudioContext()
    this.source = this.context.createMediaElementSource(element)
    this.element = element
  },

  getContext (): AudioContext {
    return this.context!
  },

  getSource (): MediaElementAudioSourceNode {
    return this.source!
  },

  getElement (): HTMLMediaElement {
    return this.element!
  }
}
