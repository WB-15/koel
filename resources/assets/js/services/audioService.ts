export const audioService = {
  context: null as unknown as AudioContext,
  source: null as unknown as MediaElementAudioSourceNode,
  element: null as unknown as HTMLMediaElement,

  init (element: HTMLMediaElement) {
    const AudioContext = window.AudioContext ||
      window.webkitAudioContext ||
      window.mozAudioContext ||
      window.oAudioContext ||
      window.msAudioContext

    this.context = new AudioContext()
    this.source = this.context.createMediaElementSource(element)
    this.element = element
  },

  getContext () {
    return this.context
  },

  getSource () {
    return this.source
  },

  getElement () {
    return this.element
  }
}
