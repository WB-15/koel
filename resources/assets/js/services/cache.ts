const CACHE_EXPIRATION_TIME = 1000 * 60 * 60 * 24 // 1 day

export const Cache = {
  storage: new Map<any, {
    time: number,
    value: any
  }>(),

  normalizeKey: (key: any) => typeof key === 'object' ? JSON.stringify(key) : key,

  has (key: any) {
    return this.hit(this.normalizeKey(key))
  },

  set (key: any, value: any) {
    this.storage.set(this.normalizeKey(key), {
      value,
      time: Date.now()
    })
  },

  hit (key: any) {
    return !this.miss(this.normalizeKey(key))
  },

  miss (key: any) {
    key = this.normalizeKey(key)

    if (!this.storage.has(key)) return true
    const { time } = this.storage.get(key)!

    if (time < Date.now() - CACHE_EXPIRATION_TIME) {
      this.storage.delete(key)
      return true
    }

    return false
  },

  invalidate (key: any) {
    this.storage.delete(this.normalizeKey(key))
  },

  resolve<T> (key: any, resolver: Closure) {
    key = this.normalizeKey(key)

    this.hit(key) || this.set(key, resolver())
    return this.storage.get(key)!.value
  }
}
