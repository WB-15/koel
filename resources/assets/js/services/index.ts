export * from './httpService'
export * from './downloadService'
export * from './localStorageService'
export * from './playbackService'
export * from './youTubeService'
export * from './socketService'
export * from './audioService'
export * from './uploadService'
export * from './authService'
