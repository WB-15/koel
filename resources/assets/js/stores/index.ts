export * from './albumStore'
export * from './artistStore'
export * from './commonStore'
export * from './equalizerStore'
export * from './favoriteStore'
export * from './overviewStore'
export * from './playlistStore'
export * from './preferenceStore'
export * from './queueStore'
export * from './recentlyPlayedStore'
export * from './searchStore'
export * from './settingStore'
export * from './songStore'
export * from './themeStore'
export * from './userStore'
