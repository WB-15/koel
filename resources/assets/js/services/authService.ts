import { localStorageService } from '@/services'

const STORAGE_KEY = 'api-token'

export const authService = {
  getToken: () => localStorageService.get<string | null>(STORAGE_KEY),

  hasToken () {
    return Boolean(this.getToken())
  },

  setToken: (token: string) => localStorageService.set(STORAGE_KEY, token),
  destroy: () => localStorageService.remove(STORAGE_KEY)
}
