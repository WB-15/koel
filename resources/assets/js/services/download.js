import $ from 'jquery';
import { map } from 'lodash';

import { playlistStore, favoriteStore } from '../stores';
import { ls } from '.';

export const download = {
    /**
     * Download individual song(s).
     *
     * @param {Array.<Object>|Object} songs
     */
    fromSongs(songs) {
        songs = [].concat(songs);
        const ids = map(songs, 'id');
        const params = $.param({ songs: ids });

        return this.trigger(`songs?${params}`);
    },

    /**
     * Download all songs in an album.
     *
     * @param {Object} album
     */
    fromAlbum(album) {
        return this.trigger(`album/${album.id}`);
    },

    /**
     * Download all songs performed by an artist.
     *
     * @param {Object} artist
     */
    fromArtist(artist) {
        // It's safe to assume an artist always has songs.
        // After all, what's an artist without her songs?
        // (See what I did there? Yes, I'm advocating for women's rights).
        return this.trigger(`artist/${artist.id}`);
    },

    /**
     * Download all songs in a playlist.
     *
     * @param {Object} playlist
     */
    fromPlaylist(playlist) {
        if (!playlistStore.getSongs(playlist).length) {
            console.warn('Empty playlist.');
            return;
        }

        return this.trigger(`playlist/${playlist.id}`);
    },

    /**
     * Download all favorite songs.
     */
    fromFavorites() {
        if (!favoriteStore.all.length) {
            console.warn("You don't like any song? Come on, don't be that grumpy.");
            return;
        }

        return this.trigger('favorites');
    },

    /**
     * Build a download link using a segment and trigger it.
     *
     * @param  {string} uri The uri segment, corresponding to the song(s),
     *                      artist, playlist, or album.
     */
    trigger(uri) {
        const sep = uri.indexOf('?') === -1 ? '?' : '&';
        const frameId = `downloader${Date.now()}`;
        $(`<iframe id="${frameId}" style="display:none"></iframe`).appendTo('body');
        document.getElementById(frameId).src = `/api/download/${uri}${sep}jwt-token=${ls.get('jwt-token')}`;
    },
}
