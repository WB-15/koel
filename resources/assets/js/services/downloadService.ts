import { favoriteStore, playlistStore } from '@/stores'
import { authService } from '.'
import { alerts, arrayify } from '@/utils'

export const downloadService = {
  fromSongs (songs: Song | Song[]): void {
    const query = arrayify(songs).reduce((q, song) => `songs[]=${song.id}&${q}`, '')
    this.trigger(`songs?${query}`)
  },

  fromAlbum (album: Album): void {
    this.trigger(`album/${album.id}`)
  },

  fromArtist (artist: Artist): void {
    this.trigger(`artist/${artist.id}`)
  },

  fromPlaylist (playlist: Playlist): void {
    if (playlistStore.getSongs(playlist).length) {
      this.trigger(`playlist/${playlist.id}`)
    }
  },

  fromFavorites (): void {
    if (favoriteStore.all.length) {
      this.trigger('favorites')
    }
  },

  /**
   * Build a download link using a segment and trigger it.
   *
   * @param  {string} uri The uri segment, corresponding to the song(s),
   *                      artist, playlist, or album.
   */
  trigger: (uri: string) => {
    const sep = uri.includes('?') ? '&' : '?'
    const url = `${window.BASE_URL}download/${uri}${sep}api_token=${authService.getToken()}`

    if (KOEL_ENV === 'app') {
      require('electron').ipcRenderer.send('DOWNLOAD', url)
      alerts.success('Download started!')
    } else {
      const iframe = document.createElement('iframe')
      iframe.style.display = 'none'
      iframe.setAttribute('src', url)
      document.body.appendChild(iframe)
    }
  }
}
