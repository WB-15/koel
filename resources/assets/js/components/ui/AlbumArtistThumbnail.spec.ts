import { orderBy } from 'lodash'
import ComponentTestCase from '@/__tests__/ComponentTestCase'
import { expect, it } from 'vitest'
import { fireEvent } from '@testing-library/vue'
import { playbackService } from '@/services'
import { queueStore } from '@/stores'
import factory from '@/__tests__/factory'
import Thumbnail from './AlbumArtistThumbnail.vue'

let album: Album
let artist: Artist

new class extends ComponentTestCase {
  private renderForAlbum () {
    album = factory<Album>('album', {
      name: 'IV',
      cover: 'https://localhost/album.jpg'
    })

    return this.render(Thumbnail, {
      props: {
        entity: album
      }
    })
  }

  private renderForArtist () {
    artist = factory<Artist>('artist', {
      name: 'Led Zeppelin',
      image: 'https://localhost/blimp.jpg'
    })

    return this.render(Thumbnail, {
      props: {
        entity: artist
      }
    })
  }

  protected test () {
    it('renders for album', () => {
      expect(this.renderForAlbum().html()).toMatchSnapshot()
    })

    it('renders for artist', () => {
      expect(this.renderForArtist().html()).toMatchSnapshot()
    })

    it('plays album', async () => {
      const mock = this.mock(playbackService, 'playAllInAlbum')
      const { getByRole } = this.renderForAlbum()

      await fireEvent.click(getByRole('button'))

      expect(mock).toHaveBeenCalledWith(album, false)
    })

    it('queues album', async () => {
      const mock = this.mock(queueStore, 'queue')
      const { getByRole } = this.renderForAlbum()

      await fireEvent.click(getByRole('button'), { altKey: true })

      expect(mock).toHaveBeenCalledWith(orderBy(album.songs, ['disc', 'track']))
    })

    it('plays artist', async () => {
      const mock = this.mock(playbackService, 'playAllByArtist')
      const { getByRole } = this.renderForArtist()

      await fireEvent.click(getByRole('button'))

      expect(mock).toHaveBeenCalledWith(artist, false)
    })

    it('queues artist', async () => {
      const mock = this.mock(queueStore, 'queue')
      const { getByRole } = this.renderForArtist()

      await fireEvent.click(getByRole('button'), { altKey: true })

      expect(mock).toHaveBeenCalledWith(orderBy(artist.songs, ['album_id', 'disc', 'track']))
    })
  }
}
