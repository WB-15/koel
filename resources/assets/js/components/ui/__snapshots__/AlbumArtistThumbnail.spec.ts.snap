// Vitest Snapshot v1

exports[`renders for album 1`] = `<span class="cover" data-testid="album-artist-thumbnail" data-v-901ba52c=""><a class="control control-play" href="" role="button" data-v-901ba52c=""><span class="hidden" data-v-901ba52c="">Play all songs in the album IV</span><span class="icon" data-v-901ba52c=""></span></a></span>`;

exports[`renders for artist 1`] = `<span class="cover" data-testid="album-artist-thumbnail" data-v-901ba52c=""><a class="control control-play" href="" role="button" data-v-901ba52c=""><span class="hidden" data-v-901ba52c="">Play all songs by Led Zeppelin</span><span class="icon" data-v-901ba52c=""></span></a></span>`;
